neo/Consensus/ConsensusService.cs

```diff
@@ -1,659 +1,673 @@
+                    shouldSendRecovery = true;
+                    break;
+                }
+
+                if (!shouldSendRecovery) return;
+            }
+            Log($"send recovery: view={context.ViewNumber}");
+            localNode.Tell(new LocalNode.SendDirectly { Inventory = context.MakeRecoveryMessage() });
+        }
+
+        private void OnPrepareRequestReceived(ConsensusPayload payload, PrepareRequest message)
+        {
+            if (context.RequestSentOrReceived || context.NotAcceptingPayloadsDueToViewChanging) return;
+            if (payload.ValidatorIndex != context.Block.ConsensusData.PrimaryIndex || message.ViewNumber != context.ViewNumber) return;
+            Log($"{nameof(OnPrepareRequestReceived)}: height={payload.BlockIndex} view={message.ViewNumber} index={payload.ValidatorIndex} tx={message.TransactionHashes.Length}");
+            if (message.Timestamp <= context.PrevHeader.Timestamp || message.Timestamp > TimeProvider.Current.UtcNow.AddMinutes(10).ToTimestampMS())
+            {
+                Log($"Timestamp incorrect: {message.Timestamp}", LogLevel.Warning);
+                return;
+            }
+            if (message.TransactionHashes.Any(p => context.Snapshot.ContainsTransaction(p)))
+            {
+                Log($"Invalid request: transaction already exists", LogLevel.Warning);
+                return;
+            }
+
+            // Timeout extension: prepare request has been received with success
+            // around 2*15/M=30.0/5 ~ 40% block time (for M=5)
+            ExtendTimerByFactor(2);
+
+            context.Block.Timestamp = message.Timestamp;
+            context.Block.ConsensusData.Nonce = message.Nonce;
+            context.TransactionHashes = message.TransactionHashes;
+            context.Transactions = new Dictionary<UInt256, Transaction>();
+            for (int i = 0; i < context.PreparationPayloads.Length; i++)
+                if (context.PreparationPayloads[i] != null)
+                    if (!context.PreparationPayloads[i].GetDeserializedMessage<PrepareResponse>().PreparationHash.Equals(payload.Hash))
+                        context.PreparationPayloads[i] = null;
+            context.PreparationPayloads[payload.ValidatorIndex] = payload;
+            byte[] hashData = context.EnsureHeader().GetHashData();
+            for (int i = 0; i < context.CommitPayloads.Length; i++)
+                if (context.CommitPayloads[i]?.ConsensusMessage.ViewNumber == context.ViewNumber)
+                    if (!Crypto.Default.VerifySignature(hashData, context.CommitPayloads[i].GetDeserializedMessage<Commit>().Signature, context.Validators[i].EncodePoint(false)))
+                        context.CommitPayloads[i] = null;
+
+            if (context.TransactionHashes.Length == 0)
+            {
+                // There are no tx so we should act like if all the transactions was filled
+                CheckPrepareResponse();

```

Comment:
Is this producing blocks in less than 15 seconds when there are no transactions?

Reply:
Forget it, I got confused. Checked that times are working fine.